false positive super-init-not-called for subclasses of typing.Protocol

### Bug description

Given the example code below, both `MyProtocol.__init__()` and `ExplicitProtocolImplimentation.__init__()` are being flagged with `super-init-not-called`.

```python
from abc import abstractmethod
from typing import Any, Protocol


class MyProtocol(Protocol):
    @abstractmethod
    def __init__(self, foo: Any) -> None:
        raise NotImplementedError


class ExplicitProtocolImplimentation(MyProtocol):
    def __init__(self, foo: Any) -> None:
        print(foo)
```
The results are the same with/without using `@abstractmethod`
and/or replacing `typing.Protocol` with `typing_extensions.Protocol`.

### Configuration

_No response_

### Command used

```shell
pylint test.py
```


### Pylint output

```shell
test.py:7:4: W0231: __init__ method from base class 'Protocol' is not called (super-init-not-called)
test.py:12:4: W0231: __init__ method from base class 'MyProtocol' is not called (super-init-not-called)
```


### Expected behavior

I would expect that neither of these would be flagged with `super-init-not-called`.

While calling `super().__init__()` here won't cause a runtime error, it also won't do anything. As noted in [PEP 544](https://www.python.org/dev/peps/pep-0544/), _[Protocols cannot be instantiated](https://www.python.org/dev/peps/pep-0544/#subtyping-relationships-with-other-types)_. This is enforced by the removal of the implementation of `MyProtocol.__init__()` at runtime.

I would also expect that trying to use `super().__init__()` in `MyProtocol.__init__()` or `ExplicitProtocolImplimentation.__init__()` would result in a warning since we can deduce that it is unnecessary (e.g. `superfluous-super`) but this could be addressed in a separate issue as a feature request.

### Pylint version

```shell
pylint 2.9.6
astroid 2.6.5
Python 3.9.6 (default, Jul 15 2021, 14:52:44) 
[GCC 9.3.0]
```


### OS / Environment

_No response_

### Additional dependencies

_No response_

--- a/ChangeLog
+++ b/ChangeLog
@@ -206,6 +206,10 @@ Release date: TBA
 * The ``PyLinter`` class will now be initialized with a ``TextReporter``
   as its reporter if none is provided.
 
+* Fix ``super-init-not-called`` when parent or ``self`` is a ``Protocol``
+
+  Closes #4790
+
 * Fix false positive ``not-callable`` with attributes that alias ``NamedTuple``
 
   Partially closes #1730

--- a/doc/whatsnew/2.13.rst
+++ b/doc/whatsnew/2.13.rst
@@ -111,6 +111,10 @@ Other Changes
 
 .. _`emacs file locks`: https://www.gnu.org/software/emacs/manual/html_node/elisp/File-Locks.html
 
+* Fix ``super-init-not-called`` when parent or ``self`` is a ``Protocol``
+
+  Closes #4790
+
 * An astroid issue where symlinks were not being taken into account
   was fixed
 

--- a/pylint/checkers/classes/class_checker.py
+++ b/pylint/checkers/classes/class_checker.py
@@ -94,7 +94,6 @@ ASTROID_TYPE_COMPARATORS = {
     nodes.Dict: lambda a, b: a.items == b.items,
     nodes.Name: lambda a, b: set(a.infer()) == set(b.infer()),
 }
-PROTOCOLS = {"typing.Protocol", "typing_extensions.Protocol"}
 
 # Dealing with useless override detection, with regard
 # to parameters vs arguments
@@ -1982,14 +1981,18 @@ a metaclass class method.",
                 continue
         for klass, method in not_called_yet.items():
             # Return if klass is protocol
-            if klass.qname() in PROTOCOLS:
+            if klass.qname() in utils.TYPING_PROTOCOLS:
                 return
 
             # Return if any of the klass' first-order bases is protocol
             for base in klass.bases:
-                inferred_base = safe_infer(base)
-                if inferred_base and inferred_base.qname() in PROTOCOLS:
-                    return
+                try:
+                    inferred_base = base.infer()
+                except astroid.InferenceError:
+                    continue
+                for inf_base in inferred_base:
+                    if inf_base.qname() in utils.TYPING_PROTOCOLS:
+                        return
 
             if decorated_with(node, ["typing.overload"]):
                 continue

--- a/pylint/checkers/classes/class_checker.py
+++ b/pylint/checkers/classes/class_checker.py
@@ -1986,11 +1986,9 @@ a metaclass class method.",
 
             # Return if any of the klass' first-order bases is protocol
             for base in klass.bases:
-                try:
-                    inferred_base = base.infer()
-                except astroid.InferenceError:
-                    continue
-                for inf_base in inferred_base:
+                # We don't need to catch InferenceError here as _ancestors_to_call
+                # already does this for us.
+                for inf_base in base.infer():
                     if inf_base.qname() in utils.TYPING_PROTOCOLS:
                         return
 

--- a/pylint/checkers/classes/class_checker.py
+++ b/pylint/checkers/classes/class_checker.py
@@ -52,10 +52,10 @@
 """Classes checker for Python code"""
 import collections
 from itertools import chain, zip_longest
-from typing import List, Pattern
+from typing import Dict, List, Pattern
 
 import astroid
-from astroid import nodes
+from astroid import bases, nodes
 
 from pylint.checkers import BaseChecker, utils
 from pylint.checkers.utils import (
@@ -81,7 +81,7 @@ from pylint.checkers.utils import (
     unimplemented_abstract_methods,
     uninferable_final_decorators,
 )
-from pylint.interfaces import IAstroidChecker
+from pylint.interfaces import INFERENCE, IAstroidChecker
 from pylint.utils import get_global_option
 
 INVALID_BASE_CLASSES = {"bool", "range", "slice", "memoryview"}
@@ -94,6 +94,7 @@ ASTROID_TYPE_COMPARATORS = {
     nodes.Dict: lambda a, b: a.items == b.items,
     nodes.Name: lambda a, b: set(a.infer()) == set(b.infer()),
 }
+PROTOCOLS = {"typing.Protocol", "typing_extensions.Protocol"}
 
 # Dealing with useless override detection, with regard
 # to parameters vs arguments
@@ -1086,12 +1087,12 @@ a metaclass class method.",
         self._check_useless_super_delegation(node)
         self._check_property_with_parameters(node)
 
-        klass = node.parent.frame(future=True)
+        klass = node.parent.frame(future=True)  # type: nodes.ClassDef # See is_method()
         self._meth_could_be_func = True
         # check first argument is self if this is actually a method
         self._check_first_arg_for_type(node, klass.type == "metaclass")
         if node.name == "__init__":
-            self._check_init(node)
+            self._check_init(node, klass)
             return
         # check signature if the method overloads inherited method
         for overridden in klass.local_attr_ancestors(node.name):
@@ -1928,7 +1929,7 @@ a metaclass class method.",
                 continue
             self.add_message("abstract-method", node=node, args=(name, owner.name))
 
-    def _check_init(self, node):
+    def _check_init(self, node: nodes.FunctionDef, klass_node: nodes.ClassDef) -> None:
         """check that the __init__ method call super or ancestors'__init__
         method (unless it is used for type hinting with `typing.overload`)
         """
@@ -1936,7 +1937,6 @@ a metaclass class method.",
             "super-init-not-called"
         ) and not self.linter.is_message_enabled("non-parent-init-called"):
             return
-        klass_node = node.parent.frame(future=True)
         to_call = _ancestors_to_call(klass_node)
         not_called_yet = dict(to_call)
         for stmt in node.nodes_of_class(nodes.Call):
@@ -1980,12 +1980,27 @@ a metaclass class method.",
             except astroid.InferenceError:
                 continue
         for klass, method in not_called_yet.items():
+            # Return if klass is protocol
+            if klass.qname() in PROTOCOLS:
+                return
+
+            # Return if any of the klass' first-order bases is protocol
+            for base in klass.bases:
+                inferred_base = safe_infer(base)
+                if inferred_base and inferred_base.qname() in PROTOCOLS:
+                    return
+
             if decorated_with(node, ["typing.overload"]):
                 continue
             cls = node_frame_class(method)
             if klass.name == "object" or (cls and cls.name == "object"):
                 continue
-            self.add_message("super-init-not-called", args=klass.name, node=node)
+            self.add_message(
+                "super-init-not-called",
+                args=klass.name,
+                node=node,
+                confidence=INFERENCE,
+            )
 
     def _check_signature(self, method1, refmethod, class_type, cls):
         """check that the signature of the two given methods match"""
@@ -2092,11 +2107,13 @@ a metaclass class method.",
         return isinstance(node, nodes.Name) and node.name == first_attr
 
 
-def _ancestors_to_call(klass_node, method="__init__"):
+def _ancestors_to_call(
+    klass_node: nodes.ClassDef, method="__init__"
+) -> Dict[nodes.ClassDef, bases.UnboundMethod]:
     """return a dictionary where keys are the list of base classes providing
     the queried method, and so that should/may be called from the method node
     """
-    to_call = {}
+    to_call: Dict[nodes.ClassDef, bases.UnboundMethod] = {}
     for base_node in klass_node.ancestors(recurs=False):
         try:
             to_call[base_node] = next(base_node.igetattr(method))

--- a/tests/functional/i/init_not_called.txt
+++ b/tests/functional/i/init_not_called.txt
@@ -1,2 +1,2 @@
-super-init-not-called:25:4:26:27:ZZZZ.__init__:__init__ method from base class 'BBBB' is not called:UNDEFINED
-super-init-not-called:58:4:59:20:AssignedInit.__init__:__init__ method from base class 'NewStyleC' is not called:UNDEFINED
+super-init-not-called:25:4:26:27:ZZZZ.__init__:__init__ method from base class 'BBBB' is not called:INFERENCE
+super-init-not-called:58:4:59:20:AssignedInit.__init__:__init__ method from base class 'NewStyleC' is not called:INFERENCE

--- a/tests/functional/n/non/non_init_parent_called.txt
+++ b/tests/functional/n/non/non_init_parent_called.txt
@@ -2,5 +2,5 @@ import-error:7:0:7:18::Unable to import 'nonexistant':UNDEFINED
 non-parent-init-called:15:8:15:26:AAAA.__init__:__init__ method from a non direct base class 'BBBBMixin' is called:UNDEFINED
 no-member:23:50:23:77:CCC:Module 'functional.n.non.non_init_parent_called' has no 'BBBB' member:INFERENCE
 no-member:28:8:28:35:CCC.__init__:Module 'functional.n.non.non_init_parent_called' has no 'BBBB' member:INFERENCE
-super-init-not-called:49:4:51:25:Super2.__init__:__init__ method from base class 'dict' is not called:UNDEFINED
+super-init-not-called:49:4:51:25:Super2.__init__:__init__ method from base class 'dict' is not called:INFERENCE
 no-member:51:8:51:23:Super2.__init__:Super of 'Super2' has no '__woohoo__' member:INFERENCE

--- a/tests/functional/s/super/super_init_not_called.py
+++ b/tests/functional/s/super/super_init_not_called.py
@@ -3,8 +3,6 @@
 
 import ctypes
 
-from typing_extensions import Protocol as ExtensionProtocol
-
 
 class Foo(ctypes.BigEndianStructure):
     """This class should not emit a super-init-not-called warning.
@@ -15,24 +13,6 @@ class Foo(ctypes.BigEndianStructure):
 
     def __init__(self):
         ctypes.BigEndianStructure.__init__(self)
-
-
-class TestProto(ExtensionProtocol):
-    """A protocol without __init__ using Protocol from typing_extensions."""
-
-
-class TestParent(TestProto):
-    """An implementation."""
-
-    def __init__(self):
-        ...
-
-
-class TestChild(TestParent):
-    """An implementation which should call the init of TestParent."""
-
-    def __init__(self):  # [super-init-not-called]
-        ...
 
 
 class UninferableChild(UninferableParent):  # [undefined-variable]

--- a/tests/functional/s/super/super_init_not_called.py
+++ b/tests/functional/s/super/super_init_not_called.py
@@ -2,8 +2,6 @@
 # pylint: disable=too-few-public-methods
 
 import ctypes
-from abc import abstractmethod
-from typing import Protocol
 
 from typing_extensions import Protocol as ExtensionProtocol
 
@@ -17,21 +15,6 @@ class Foo(ctypes.BigEndianStructure):
 
     def __init__(self):
         ctypes.BigEndianStructure.__init__(self)
-
-
-class MyProtocol(Protocol):
-    """A protocol."""
-
-    @abstractmethod
-    def __init__(self) -> None:
-        raise NotImplementedError
-
-
-class ProtocolImplimentation(MyProtocol):
-    """An implementation."""
-
-    def __init__(self) -> None:
-        ...
 
 
 class TestProto(ExtensionProtocol):

--- a/tests/functional/s/super/super_init_not_called.py
+++ b/tests/functional/s/super/super_init_not_called.py
@@ -29,6 +29,13 @@ class TestParent(TestProto):
 
 
 class TestChild(TestParent):
+    """An implementation which should call the init of TestParent."""
+
+    def __init__(self):  # [super-init-not-called]
+        ...
+
+
+class UninferableChild(UninferableParent):
     """An implementation which should could the init of TestParent."""
 
     def __init__(self):  # [super-init-not-called]

--- a/tests/functional/s/super/super_init_not_called.py
+++ b/tests/functional/s/super/super_init_not_called.py
@@ -35,8 +35,8 @@ class TestChild(TestParent):
         ...
 
 
-class UninferableChild(UninferableParent):
-    """An implementation which should could the init of TestParent."""
+class UninferableChild(UninferableParent):  # [undefined-variable]
+    """An implementation that test if we don't crash on uninferable parents."""
 
-    def __init__(self):  # [super-init-not-called]
+    def __init__(self):
         ...

--- a/tests/functional/s/super/super_init_not_called.py
+++ b/tests/functional/s/super/super_init_not_called.py
@@ -1,13 +1,51 @@
-"""This should not emit a super-init-not-called warning. It previously did this, because
-``next(node.infer())`` was used in that checker's logic and the first inferred node
-was an Uninferable object, leading to this false positive."""
+"""Tests for super-init-not-called."""
 # pylint: disable=too-few-public-methods
 
 import ctypes
+from abc import abstractmethod
+from typing import Protocol
+
+from typing_extensions import Protocol as ExtensionProtocol
 
 
+# This should not emit a super-init-not-called warning. It previously did this, because
+# ``next(node.infer())`` was used in that checker's logic and the first inferred node
+# was an Uninferable object, leading to this false positive.
 class Foo(ctypes.BigEndianStructure):
     """A class"""
 
     def __init__(self):
         ctypes.BigEndianStructure.__init__(self)
+
+
+class MyProtocol(Protocol):
+    """A protocol."""
+
+    @abstractmethod
+    def __init__(self) -> None:
+        raise NotImplementedError
+
+
+class ProtocolImplimentation(MyProtocol):
+    """An implementation."""
+
+    def __init__(self) -> None:
+        ...
+
+
+class TestProto(ExtensionProtocol):
+    """A protocol without __init__ using Protocol from typing_extensions."""
+
+
+class TestParent(TestProto):
+    """An implementation."""
+
+    def __init__(self):
+        ...
+
+
+class TestChild(TestParent):
+    """An implementation which should could the init of TestParent."""
+
+    def __init__(self):  # [super-init-not-called]
+        ...

--- a/tests/functional/s/super/super_init_not_called.txt
+++ b/tests/functional/s/super/super_init_not_called.txt
@@ -1,1 +1,1 @@
-super-init-not-called:51:4:52:11:TestChild.__init__:__init__ method from base class 'TestParent' is not called:INFERENCE
+super-init-not-called:34:4:35:11:TestChild.__init__:__init__ method from base class 'TestParent' is not called:INFERENCE

--- a/tests/functional/s/super/super_init_not_called.txt
+++ b/tests/functional/s/super/super_init_not_called.txt
@@ -1,0 +1,1 @@
+super-init-not-called:50:4:51:11:TestChild.__init__:__init__ method from base class 'TestParent' is not called:INFERENCE

--- a/tests/functional/s/super/super_init_not_called.txt
+++ b/tests/functional/s/super/super_init_not_called.txt
@@ -1,1 +1,2 @@
 super-init-not-called:34:4:35:11:TestChild.__init__:__init__ method from base class 'TestParent' is not called:INFERENCE
+undefined-variable:38:23:38:40:UninferableChild:Undefined variable 'UninferableParent':UNDEFINED

--- a/tests/functional/s/super/super_init_not_called.txt
+++ b/tests/functional/s/super/super_init_not_called.txt
@@ -1,2 +1,1 @@
-super-init-not-called:34:4:35:11:TestChild.__init__:__init__ method from base class 'TestParent' is not called:INFERENCE
-undefined-variable:38:23:38:40:UninferableChild:Undefined variable 'UninferableParent':UNDEFINED
+undefined-variable:18:23:18:40:UninferableChild:Undefined variable 'UninferableParent':UNDEFINED

--- a/tests/functional/s/super/super_init_not_called_extensions.py
+++ b/tests/functional/s/super/super_init_not_called_extensions.py
@@ -1,0 +1,22 @@
+"""Tests for super-init-not-called."""
+# pylint: disable=too-few-public-methods
+
+from typing_extensions import Protocol as ExtensionProtocol
+
+
+class TestProto(ExtensionProtocol):
+    """A protocol without __init__ using Protocol from typing_extensions."""
+
+
+class TestParent(TestProto):
+    """An implementation."""
+
+    def __init__(self):
+        ...
+
+
+class TestChild(TestParent):
+    """An implementation which should call the init of TestParent."""
+
+    def __init__(self):  # [super-init-not-called]
+        ...

--- a/tests/functional/s/super/super_init_not_called_extensions.rc
+++ b/tests/functional/s/super/super_init_not_called_extensions.rc
@@ -1,0 +1,2 @@
+[testoptions]
+max_pyver=3.9

--- a/tests/functional/s/super/super_init_not_called_extensions.txt
+++ b/tests/functional/s/super/super_init_not_called_extensions.txt
@@ -1,0 +1,1 @@
+super-init-not-called:21:4:22:11:TestChild.__init__:__init__ method from base class 'TestParent' is not called:INFERENCE

--- a/tests/functional/s/super/super_init_not_called_extensions_py310.py
+++ b/tests/functional/s/super/super_init_not_called_extensions_py310.py
@@ -1,0 +1,22 @@
+"""Tests for super-init-not-called."""
+# pylint: disable=too-few-public-methods
+
+from typing_extensions import Protocol as ExtensionProtocol
+
+
+class TestProto(ExtensionProtocol):
+    """A protocol without __init__ using Protocol from typing_extensions."""
+
+
+class TestParent(TestProto):
+    """An implementation."""
+
+    def __init__(self):
+        ...
+
+
+class TestChild(TestParent):
+    """An implementation which should call the init of TestParent."""
+
+    def __init__(self):  # [super-init-not-called]
+        ...

--- a/tests/functional/s/super/super_init_not_called_extensions_py310.rc
+++ b/tests/functional/s/super/super_init_not_called_extensions_py310.rc
@@ -1,0 +1,4 @@
+[testoptions]
+# Windows test environments on >= 3.10 don't have typing_extensions
+exclude_platforms=win32
+min_pyver=3.10

--- a/tests/functional/s/super/super_init_not_called_extensions_py310.txt
+++ b/tests/functional/s/super/super_init_not_called_extensions_py310.txt
@@ -1,0 +1,1 @@
+super-init-not-called:21:4:22:11:TestChild.__init__:__init__ method from base class 'TestParent' is not called:INFERENCE

--- a/tests/functional/s/super/super_init_not_called_py38.py
+++ b/tests/functional/s/super/super_init_not_called_py38.py
@@ -1,0 +1,20 @@
+"""Tests for super-init-not-called with Protocol."""
+# pylint: disable=too-few-public-methods
+
+from abc import abstractmethod
+from typing import Protocol
+
+
+class MyProtocol(Protocol):
+    """A protocol."""
+
+    @abstractmethod
+    def __init__(self) -> None:
+        raise NotImplementedError
+
+
+class ProtocolImplimentation(MyProtocol):
+    """An implementation."""
+
+    def __init__(self) -> None:
+        ...

--- a/tests/functional/s/super/super_init_not_called_py38.rc
+++ b/tests/functional/s/super/super_init_not_called_py38.rc
@@ -1,0 +1,2 @@
+[testoptions]
+min_pyver=3.8

